### DISCUSSION TITLE
How to efficiently debug localization ?

### DISCUSSION THREAD

**Discussion by mattiasmar** [introlab/rtabmap#1293] [How to]
With the rtabmap-databaseviewer it is easy to review the results of mapping.
What is the recommended way to debug rtabmap for localization? Can I use the databaseviewer for that? What I think I would like to have is an option to load the databaseviewer with 2 rtabmap.db files; the original mapping db and a kind of db file that would be generated from running rtabmap in localization mode.
@matlabbe  Is there an option for this? How do you recommend to debug localization mode? How can one generate a rtabmap.db file that holds the images of the localization mode only?

**Comment by matlabbe** [comment]
Using two databases like this is the way I debugged localization and got localization results for that [paper](https://github.com/introlab/rtabmap/tree/master/archive/2022-IlluminationInvariant). The trick is to enable the parameter `Mem/LocalizationDataSaved` to save images/statistics that occurred during the localization session. The localization data in the database will be disjoint from the actual graph, but can be browsed (as well at the statistics) with `rtabmap-databaseViewer`.

Example:
```
rtabmap-reprocess --Mem/IncrementalMemory false --Mem/LocalizationDataSaved true \
   "original_map.db;another_map_to_test_localization.db" localization_results.db
```
This will initialize `reprocess` with the initial map `original_map.db`, then do localization session with `another_map_to_test_localization.db`. The resulting map `localization_results.db` will contain the original map, but also all statistics/data of the localization session.

**Comment by mattiasmar** [comment]
Thanks!

**Comment by mattiasmar** [comment]
I see that the rtabmap-reprocess runs in localization mode using the above command, and that's good. However, in the DB-viewer I do not see any trace of the localization. There are no more images seen than what the original mapping db held. What I would like to see, is the  registration results between the localization.db vs the mapping.db and the node links (as seen in the "Graph View") between the localization graph and the mapping graph. Is any of this possible?

**Reply by matlabbe** [reply]
With `--Mem/LocalizationDataSaved true`, the localization images should be there (the "Add" button in Constraints view can be enabled by clicking the checkbox next to it to test localization between localization frame and another one in the map). The Statistics would also contain statistics of the localization session. We cannot see the localization constraints though, the Graph View will show only the map constraints. It is however possible to recover the localization trajectory from the corresponding statistics contained in the database (e.g., `Loop/MapToBase_x/m` and `Loop/MapToBase_y/m`).
